firmware/ethereum.c

```diff
@@ -0,0 +1,449 @@
+{
+	signing = true;
+	sha3_256_Init(&keccak_ctx);
+
+	memset(&resp, 0, sizeof(EthereumTxRequest));
+
+	if (msg->has_data_length) {
+		if (msg->data_length == 0) {
+			fsm_sendFailure(FailureType_Failure_Other, "Invalid data length provided");
+			ethereum_signing_abort();
+			return;
+		}
+		if (!msg->has_data_initial_chunk || msg->data_initial_chunk.size == 0) {
+			fsm_sendFailure(FailureType_Failure_Other, "Data length provided, but no initial chunk");
+			ethereum_signing_abort();
+			return;
+		}
+		if (msg->data_initial_chunk.size > msg->data_length) {
+			fsm_sendFailure(FailureType_Failure_Other, "Invalid size of initial chunk");
+			ethereum_signing_abort();
+			return;
+		}
+		data_total = msg->data_length;
+	} else {
+		data_total = 0;
+	}
+
+	// safety checks
+	if (!ethereum_signing_check(msg)) {
+		fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing aborted (safety check failed)");
+		ethereum_signing_abort();
+		return;
+	}
+
+	layoutEthereumConfirmTx(msg->has_to ? msg->to.bytes : NULL, msg->has_value ? msg->value.bytes : NULL, msg->has_value ? msg->value.size : 0);
+	if (!protectButton(ButtonRequestType_ButtonRequest_SignTx, false)) {
+		fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled by user");
+		ethereum_signing_abort();
+		return;
+	}
+
+	/* Stage 1: Calculate total RLP length */
+	uint32_t rlp_length = 0;
+
+	layoutProgress("Signing", 0);
+
+	rlp_length += msg->has_nonce ? rlp_calculate_length(msg->nonce.size, msg->nonce.bytes[0]) : 1;
+	rlp_length += msg->has_gas_price ? rlp_calculate_length(msg->gas_price.size, msg->gas_price.bytes[0]) : 1;
+	rlp_length += msg->has_gas_limit ? rlp_calculate_length(msg->gas_limit.size, msg->gas_limit.bytes[0]) : 1;
+	rlp_length += msg->has_to ? rlp_calculate_length(msg->to.size, msg->to.bytes[0]) : 1;
+	rlp_length += msg->has_value ? rlp_calculate_length(msg->value.size, msg->value.bytes[0]) : 1;
+	rlp_length += (msg->has_data_length && msg->has_data_initial_chunk) ? rlp_calculate_length(msg->data_length, msg->data_initial_chunk.bytes[0]) : 1;
+
+	/* Stage 2: Store header fields */
+	hash_rlp_list_length(rlp_length);
+
+	layoutProgress("Signing", 100);
+
+	if (msg->has_nonce) {
+		hash_rlp_field(msg->nonce.bytes, msg->nonce.size);
+	} else {
+		hash_rlp_length(1, 0);
+	}
+
+	if (msg->has_gas_price) {
+		hash_rlp_field(msg->gas_price.bytes, msg->gas_price.size);
+	} else {
+		hash_rlp_length(1, 0);
+	}
+
+	if (msg->has_gas_limit) {
+		hash_rlp_field(msg->gas_limit.bytes, msg->gas_limit.size);
+	} else {
+		hash_rlp_length(1, 0);
+	}
+
+	if (msg->has_to) {
+		hash_rlp_field(msg->to.bytes, msg->to.size);
+	} else {
+		hash_rlp_length(1, 0);
+	}
+
+	if (msg->has_value) {
+		hash_rlp_field(msg->value.bytes, msg->value.size);
+	} else {
+		hash_rlp_length(1, 0);
+	}
+
+	if (msg->has_data_length && msg->has_data_initial_chunk) {
+		hash_rlp_length(msg->data_length, msg->data_initial_chunk.bytes[0]);
```

Comment:
This should use the value data_total instead of msg->data-length.  Also the if is unnecessary; if data_total is 0 (and therefore, also data_initial_chunk_size) the then part does the same as the else part.